### PR TITLE
fix(auth): constant-time password comparison

### PR DESCRIPTION
**Pull request by TickTockBent**
## Description

Fixes #177.

Replaces three `===` password comparisons with a constant-time helper so an
attacker with timing access cannot statistically recover the password byte by
byte.

Sites changed:

- `src/server/routes.js:136` - `POST /api/auth` login compare
- `src/server/websocket.js:129` - WS `auth` message compare
- `src/server/auth.js:376` - `Authorization: Bearer <pw>` compare (same bug
  class, not in the original report but fixed opportunistically since the
  helper was already landing)

### Approach

Added `safeCompare(a, b)` in `src/server/auth.js`. It SHA-256 hashes both
inputs and runs `crypto.timingSafeEqual` on the digests. Hashing first means:

- equal-length buffers every time, so `timingSafeEqual` never throws on a
  length mismatch
- no length leak from the comparison itself
- non-string inputs return `false` via an early `typeof` guard without
  touching the hash path

Exposed on the `createAuth` return object so call sites use `auth.safeCompare`
through the same dependency they already hold. Also exported at module level
so the WebSocket test mock reuses the real implementation.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactor / chore (no functional changes)

## Checklist

- [x] My code follows the existing code style
- [x] I have tested my changes locally
- [ ] I have tested on mobile (if UI changes) - n/a, server-only change
- [x] Tests pass (`npm test`)
- [ ] I have updated documentation (if needed) - n/a
- [x] My commits follow [conventional commits](https://www.conventionalcommits.org/)

## Related Issues

Fixes #177.

---

### Test notes

New `safeCompare` suite in `test/server/auth.test.js` covers: identical
strings, different strings, length mismatch, non-string inputs
(`null`/`undefined`/number/object), unicode, and the empty-string edge case.
Mock auth in `test/server/websocket.test.js` now exposes `safeCompare` so WS
auth tests exercise the new code path.

`npm test` passes the full auth + websocket suites (107 tests). 2 pre-existing
failures in `routes.test.js` (SPA `index.html` 404s) reproduce on clean `main`
and are unrelated to this change.

// ticktockbent

### CODE REVIEW

File: src/server/auth.js [hunk 1]
```diff
@@ -290,6 +290,15 @@
 </body>
 </html>`;
 
+// Constant-time string compare. Hashing both sides first masks length
+// differences and guarantees equal-length buffers for timingSafeEqual.
+function safeCompare(a, b) {
+  if (typeof a !== 'string' || typeof b !== 'string') return false;
+  const ah = crypto.createHash('sha256').update(a).digest();
+  const bh = crypto.createHash('sha256').update(b).digest();
```
[the remark this answers is not in the document]
Comment (dorlugasigal) [reply src/server/auth.js]: Addressed in f2c46c6: switched from plain SHA-256 to HMAC-SHA256 keyed with a per-process random secret. Functionally equivalent for our use (ephemeral equality, never stored), and HMAC with a secret key is the accepted primitive for constant-time compare — should satisfy the rule.

**Comment by dorlugasigal** [comment]
Thanks a ton for this contribution, @TickTockBent! 🙏

Shipped as part of the next patch release. Small note on follow-up iterations in case you're curious:

- CodeQL's `js/insufficient-password-hash` rule taint-tracks any hashing primitive (including HMAC) applied to a password-labeled value, so we ended up dropping the hash entirely and using `crypto.timingSafeEqual` on fixed-length zero-padded buffers (256 bytes) with a separate length check. Same constant-time guarantee, no hash primitive for CodeQL to flag.
- Also updated `docs/security.md` and the agent instructions to reflect the new behavior.

Really appreciate the patch — keep 'em coming!
